py/makeqstrdata.py

```diff
@@ -52,7 +52,8 @@
 codepoint2name[ord("|")] = "pipe"
 codepoint2name[ord("~")] = "tilde"
 
-# static qstrs, should be sorted
+# static qstrs, these must maintain a specific order for .mpy compatibility
```

Comment:
Maybe more extreme version of the same idea, but what about adding a constant lookup table like `mp_binary_op_method_name` for these? That would add 157 byte code size, but also allow this whole qstr pool to be sorted. In the next .mpy version bump the lookup table could be removed to get the code size back.

(This would also potentially allow deleting the `!sorted` case from the search lookup code, wouldn't it?)

Reply:
I like this idea! It might be easier just to go with three sorted pools (because this list is _almost_ sorted) as described above.

Unfortunately, any runtime qstrs are not sorted (because the pools are created incrementally), so the `!sorted` case needs to stay.

Reply:
I'm guessing it's not viable to insertion sort the runtime pool? I'll stop proposing ever more complex changes now, I promise. :laughing:

Reply:
@projectgus 

The main problem is that we return the qstr id at insertion time. This id is the index across all the pools (i.e. the first pool contains qstr 0 -> N, the second contains N+1 -> M, etc).

So sorting would require some way to know the final index in the pool, even though the sort ordering might change with further insertions.

There are easy ways to solve this with extra indirection of course, but that would be an extra `2*sizeof(qstr)` per entry in the runtime tables. Maybe this is worth considering?

Also other ideas around some sort of hash-table-like approach. The problem with hashtables though is that the main expensive operation we perform on the pool is set-existence (i.e. we have this new string from _somewhere_, and we're trying to decide whether it's an existing qstr or not). And worse, the expected outcome for any given pool is negative, which is pathological for high-load-factor hashtables.